File: docs/JAMIA/comparisons/RPython/TM/22481271_corr_object_dtype.py
# ID: SO#22481271
# Language: Python
# Bug class: Value type
# Runtime consequence: IF
# In study: Yes
# Inclusion rationale: corr() returns empty matrix on object-dtype columns. Numeric operation on string-typed data silently produces wrong result.
# Reproduction status: Live
# Type system catch: `Record<string, number>` rejects string column

import pandas as pd

# Columns are object dtype — values look numeric but aren't typed as such.
# This commonly happens through CSV loading, .T on mixed frames, or concat.
df = pd.DataFrame({
    "A": ["0.006", "-0.002", "0.010", "0.003", "0.002"],
    "B": ["-0.001", "-0.0005", "0.0003", "0.001", "-0.0002"],
    "C": ["0.003", "-0.002", "0.002", "-0.003", "0.002"],
})

print("dtypes:", df.dtypes.tolist())
print("Data looks numeric:")
print(df)
print()

# With numeric_only=True (the old default), silently returns empty
result = df.corr(numeric_only=True)
print("df.corr(numeric_only=True):")
print(result)
print()

# User expects a 3x3 matrix, gets nothing. No error, no warning.
# In a clinical pipeline, this would mean correlations silently missing
# from a report — the analysis appears complete but is empty.

File: docs/JAMIA/comparisons/RPython/TM/22481271_corr_object_dtype.ts
/**
 * RPython SO#22481271 — corr() on object-dtype columns
 *
 * In pandas, CSV data can arrive as object dtype (looks numeric, isn't).
 * DataFrame.corr() silently returns an empty matrix — no error.
 *
 * In tidy-ts, CSV parsing produces typed columns. If a column parsed as
 * string (due to leading spaces, mixed formats), the type system reflects
 * that. Passing string columns to s.test.correlation.pearson is a
 * compile-time error — you must explicitly convert first.
 */
import { createDataFrame, stats as s } from "@tidy-ts/dataframe";

// If CSV parsing produced string columns (e.g., leading spaces prevented
// numeric parsing), the schema reflects reality:
const df = createDataFrame([
  { A: " 0.006", B: "-0.001", C: " 0.003" },
  { A: " 0.002", B: "-0.001", C: "-0.002" },
  { A: " 0.010", B: " 0.000", C: " 0.002" },
]);

// Schema is { A: string, B: string, C: string }
// extract("A") returns string[] — pearson requires number[]
// @ts-expect-error — string[] is not assignable to number[]
const result = s.test.correlation.pearson({ x: df.extract("A"), y: df.extract("B") });

// The fix is explicit: parse the columns to numbers first.
const numeric = df.mutate({
  A: (r) => parseFloat(r.A),
  B: (r) => parseFloat(r.B),
  C: (r) => parseFloat(r.C),
});

// Now extract returns number[] — this works
const correct = s.test.correlation.pearson({
  x: numeric.extract("A"),
  y: numeric.extract("B"),
});
console.log(correct.effectSize.value); // the Pearson r coefficient
